src/window/update_loop.rs

```diff
@@ -175,38 +207,77 @@ impl UpdateLoop {
                 return Err(());
             }
             Ok(Event::AboutToWait) | Err(false) => {
-                self.should_render.update(window_wrapper.prepare_frame());
-                if self.should_render == ShouldRender::Immediately || !self.idle {
-                    // Always draw immediately for reduced latency if we have been idling
-                    if self.num_consecutive_rendered > 0
-                        && window_wrapper.vsync.uses_winit_throttling()
+                // We will also animate, but not render when frames are skipped(or very late) to reduce visual artifacts
```

Comment:
Style (and pretty unimportant): This is quite nested, so much that it becomes hard to follow what brace closes what. It might make sense to extract this into its own function `maybe_render` or the like?

Reply:
Yes, I think there are a few things that could be refactored here. I don't want to risk breaking any functionality right now though. So I think it can wait a bit and be done in a separate PR.

Reply:
Sounds fair!